grovedb/src/replication/state_sync_session.rs

```diff
@@ -431,29 +475,44 @@ impl<'db> MultiStateSyncSession<'db> {
                 let completed_path = subtree_state_sync.current_path.clone();
 
                 // Subtree is finished. We can save it.
-                if subtree_state_sync.num_processed_chunks > 0 {
-                    if let Some(prefix_data) = current_prefixes.remove(&chunk_prefix) {
+                let is_subtree_empty = subtree_state_sync.num_processed_chunks == 0;
+                if let Some(prefix_data) = current_prefixes.remove(&chunk_prefix) {
+                    if !is_subtree_empty {
                         if let Err(err) = prefix_data.restorer.finalize(grove_version) {
                             return Err(Error::InternalError(format!(
                                 "Unable to finalize Merk: {:?}",
                                 err
                             )));
                         }
-                    } else {
-                        return Err(Error::InternalError(format!(
-                            "Prefix {:?} does not exist in current_prefixes",
-                            chunk_prefix
-                        )));
                     }
+                } else {
+                    return Err(Error::InternalError(format!(
+                        "Prefix {:?} does not exist in current_prefixes",
+                        chunk_prefix
+                    )));
                 }
 
                 self.as_mut().processed_prefixes().insert(chunk_prefix);
 
+                *self.as_mut().num_processed_subtrees_in_batch() += 1;
+
                 let new_subtrees_metadata =
-                    self.discover_new_subtrees_metadata(db, &completed_path, grove_version)?;
+                    self.discover_new_subtrees_metadata(&completed_path, grove_version)?;
 
-                if let Ok(res) =
-                    self.prepare_sync_state_sessions(db, new_subtrees_metadata, grove_version)
+                if self.num_processed_subtrees_in_batch >= self.subtrees_batch_size {
+                    match self.as_mut().pending_discovered_subtrees() {
```

Comment:
this kind of match is preferred as `if let Some(existing_subtrees_metadata) = .. {} else {}`